### PR TITLE
clean event types

### PR DESCRIPTION
**Pull request by FriedrichWilken**
This pr addresses [this](https://github.com/kyma-project/kyma/issues/10814) issue:
When creating backend infrastructures we take the `filters` -> `eventType` -> `value`,  clean it up and name the backend infrastructure by this name, while leaving the `eventType` -> `value` untouched. This leads to confusion about the actual name of the backend infrastructure. To resolve this, the field `SubscriptionStatus` -> `CleanEventTypes` was added that contains the cleaned names.

Alongside some minor refactoring happened.

### CODE REVIEW

File: components/eventing-controller/pkg/handlers/nats.go [hunk 1]
```diff
@@ -90,20 +90,21 @@ func newCloudeventClient(config env.NatsConfig) (cev2.Client, error) {
 
 // SyncSubscription synchronizes the given Kyma subscription to NATS subscription.
 // note: the returned bool should be ignored now. It should act as a marker for changed subscription status.
-func (n *Nats) SyncSubscription(sub *eventingv1alpha1.Subscription, cleaner eventtype.Cleaner, _ ...interface{}) (bool, error) {
+func (n *Nats) SyncSubscription(subscription *eventingv1alpha1.Subscription, cleaner eventtype.Cleaner, _ ...interface{}) (bool, error) {
```
Comment (k15r) [inline components/eventing-controller/pkg/handlers/nats.go:93]: i prefer the old `sub` over the new `subscription`. This is in line with https://github.com/golang/go/wiki/CodeReviewComments#variable-names.
I would also prefer if you could do such name changes in a separate PR if they are on a larger scale. In this case one PR that changes all subscription function arguments to subs would be very welcome.

File: components/eventing-controller/pkg/handlers/nats.go [hunk 2]
```diff
@@ -90,20 +90,21 @@ func newCloudeventClient(config env.NatsConfig) (cev2.Client, error) {
...
 		filters = uniqueFilters.Filters
 	}
 
 	// Format logger
-	log := utils.LoggerWithSubscription(n.namedLogger(), sub)
+	log := utils.LoggerWithSubscription(n.namedLogger(), subscription)
 
-	subsConfig := eventingv1alpha1.MergeSubsConfigs(sub.Spec.Config, &n.defaultSubsConfig)
+	subscriptionConfig := eventingv1alpha1.MergeSubsConfigs(subscription.Spec.Config, &n.defaultSubsConfig)
+	subscription.Status.CleanEventTypes = make([]string, 0, len(filters))
```
Comment (k15r) [inline components/eventing-controller/pkg/handlers/nats.go:107]: why not use `subscription.Status.CleanEventTypes = []string{}`
Or create an array - if you already know the number of filters (if performance is of the essence here) - and access it using the index

File: components/eventing-controller/pkg/handlers/beb.go [hunk 3]
```diff
@@ -284,6 +287,14 @@ func (b *BEB) getSubscription(name string) (*types.Subscription, error) {
 	return bebSubscription, nil
 }
 
+func getCleanEventTypes(events types.Events) []string {
+	cleanEventTypes := make([]string, 0, len(events))
+	for _, e := range events {
+		cleanEventTypes = append(cleanEventTypes, e.Type)
```
Comment (k15r) [inline components/eventing-controller/pkg/handlers/beb.go:293]: how about:
```
for i, e := range events {
		cleanEventTypes[i] = e.Type
```

File: components/eventing-controller/api/v1alpha1/subscription_types.go [hunk 4]
```diff
@@ -184,6 +184,10 @@ type SubscriptionStatus struct {
 	// Ready defines the overall readiness status of a subscription
 	Ready bool `json:"ready"`
 
+	// CleanEventTypes defines the clean filter event types
```
Comment (k15r) [inline components/eventing-controller/api/v1alpha1/subscription_types.go:187]: CleanEventTypes defines the filter event types after cleanup for use with the configured backend

File: components/eventing-controller/reconciler/subscription/nats/reconciler.go [hunk 5]
```diff
@@ -208,16 +208,16 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
...
-		if err := r.syncSubscriptionStatus(ctx, actualSubscription, false, err.Error()); err != nil {
+		if err := r.syncSubscriptionStatus(ctx, desiredSubscription, false, err.Error()); err != nil {
 			return ctrl.Result{}, err
 		}
 		return ctrl.Result{}, err
 	}
 	log.Debug("create NATS subscriptions succeeded")
 
-	actualSubscription.Status.Config = desiredSubscription.Status.Config
+	//actualSubscription.Status.Config = desiredSubscription.Status.Config
```
Comment (k15r) [inline components/eventing-controller/reconciler/subscription/nats/reconciler.go:218]: not needed? remove it!

File: components/eventing-controller/pkg/handlers/beb.go [hunk 6]
```diff
@@ -159,6 +159,9 @@ func (b *BEB) SyncSubscription(subscription *eventingv1alpha1.Subscription, clea
 	// set the status of bebSubscription in ev2Subscription
 	statusChanged = b.setEmsSubscriptionStatus(subscription, bebSubscription) || statusChanged
 
+	// get the backend infrastructures
```
Comment (k15r) [inline components/eventing-controller/pkg/handlers/beb.go:162]: what are those backend infrastructures?

File: components/eventing-controller/pkg/handlers/nats.go (hunk 2 above)
Comment (FriedrichWilken) [reply components/eventing-controller/pkg/handlers/nats.go:107]: I think you are right about `[]string{}`.

About the array: I am not sure here, but I think an array needs a constant length when it gets declared, so dynamic 
declaration like `subscription.Status.CleanEventTypes = [len(filters)]string` is illegal. Further, the length of it would need to be declared when `subscription` gets created. Right?

File: components/eventing-controller/pkg/handlers/beb.go (hunk 3 above)
Comment (FriedrichWilken) [reply components/eventing-controller/pkg/handlers/beb.go:293]: Agree, not smart to create with size and then don't address by index.

File: components/eventing-controller/testing/safesubscriptions.go [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+import (
+	"strings"
+	"sync"
+
+	bebtypes "github.com/kyma-project/kyma/components/eventing-controller/pkg/ems/api/events/types"
+)
+
+// SafeSubscription encapsulates Subscriptions to provide mutual exclusion.
+type SafeSubscription struct {
+	mutex         *sync.Mutex
```
Comment (k15r) [inline components/eventing-controller/testing/safesubscriptions.go:12]: lets use `sync.RWMutex`

File: components/eventing-controller/testing/safesubscriptions.go [hunk 8]
```diff
@@ -9,35 +9,35 @@ import (
...
+	s.mutex.RLock()
+	defer s.mutex.RUnlock()
 	return s.subscriptions[key]
 }
 
 // DeleteSubscription deletes a Subscription via the corresponding key.
 func (s *SafeSubscription) DeleteSubscription(key string) {
-	s.mutex.Lock()
-	defer s.mutex.Unlock()
+	s.mutex.RLock()
```
Comment (k15r) [inline components/eventing-controller/testing/safesubscriptions.go:32]: That one modifies the map. so please write lock it.

File: components/eventing-controller/testing/safesubscriptions.go [hunk 9]
```diff
@@ -9,35 +9,35 @@ import (
...
+	defer s.mutex.RUnlock()
 	delete(s.subscriptions, key)
 }
 
 // DeleteSubscriptionsByName deletes all Subscriptions that contain the substring 'name' in their own name.
 func (s *SafeSubscription) DeleteSubscriptionsByName(name string) {
-	s.mutex.Lock()
-	defer s.mutex.Unlock()
+	s.mutex.RLock()
```
Comment (k15r) [inline components/eventing-controller/testing/safesubscriptions.go:39]: same as above

File: components/eventing-controller/testing/safesubscriptions.go [hunk 10]
```diff
@@ -47,7 +47,7 @@ func (s *SafeSubscription) DeleteSubscriptionsByName(name string) {
 
 // PutSubscription sets a Subscription of SafeSubscription via the corresponding key.
 func (s *SafeSubscription) PutSubscription(key string, subscription *bebtypes.Subscription) {
-	s.mutex.Lock()
-	defer s.mutex.Unlock()
+	s.mutex.RLock()
```
Comment (k15r) [inline components/eventing-controller/testing/safesubscriptions.go:50]: .. and here as well

**Review by k15r — DISMISSED** [review]

File: resources/eventing/values.yaml [hunk 11]
```diff
@@ -14,6 +14,15 @@ global:
...
       version: 2.6.5-alpine
       directory: external
+    nats_config_reloader:
+      name: nats-server-config-reloader
+      version: 0.6.2
+      directory: external/natsio
+    prometheus_nats_exporter:
+      name: prometheus-nats-exporter
+      version: 0.9.0
+      directory: external/natsio
```
Comment (marcobebway) [inline resources/eventing/values.yaml:24]: Why do we need those changes?

File: components/eventing-controller/Makefile [hunk 12]
```diff
@@ -58,18 +58,19 @@ set-up-local-env:
...
 ENVTEST_ASSETS_DIR=$(PROJECT_DIR)/testbin/$(shell uname)
-test-local: test-local-nonlocal; ## Run tests.
+test-local: copy-external-crds test-local-nonlocal; ## Run tests.
 
 setup-envtest:
 	mkdir -p ${ENVTEST_ASSETS_DIR}
 	test -f ${ENVTEST_ASSETS_DIR}/setup-envtest.sh || curl -sSLo ${ENVTEST_ASSETS_DIR}/setup-envtest.sh https://raw.githubusercontent.com/kubernetes-sigs/controller-runtime/v0.8.3/hack/setup-envtest.sh
 	source ${ENVTEST_ASSETS_DIR}/setup-envtest.sh; fetch_envtest_tools $(ENVTEST_ASSETS_DIR); setup_envtest_env $(ENVTEST_ASSETS_DIR)
 
+# ???
```
Comment (marcobebway) [inline components/eventing-controller/Makefile:69]: Can you please either remove this or make it complete?

File: components/eventing-controller/Makefile [hunk 13]
```diff
@@ -58,18 +58,19 @@ set-up-local-env:
...
 test-local-nonlocal: manifests-local generate-local setup-envtest
 	go test ./... -coverprofile cover.out
 
+# ???
```
Comment (marcobebway) [inline components/eventing-controller/Makefile:73]: Can you please either remove this or make it complete?

File: components/eventing-controller/controllers/backend/reconciler_test.go [hunk 14]
```diff
@@ -452,7 +452,7 @@ var _ = Describe("Backend Reconciliation Tests", func() {
 	When("Switching to BEB and then stopping NATS controller fails", func() {
 		It("Should mark Eventing Backend CR to not ready", func() {
 			ctx := context.Background()
-			natsSubMgr.stopErr = errors.New("I shan't stop")
+			natsSubMgr.stopErr = errors.New("I can't stop")
```
Comment (marcobebway) [inline components/eventing-controller/controllers/backend/reconciler_test.go:455]: Is this change needed?

File: components/eventing-controller/controllers/subscription/beb/reconciler_test.go [hunk 15]
```diff
@@ -99,6 +99,114 @@ var _ = Describe("Subscription Reconciliation Tests", func() {
 		testID++
 	})
 
+	When("Manipulating the clean event types", func() {
```
Comment (marcobebway) [inline components/eventing-controller/controllers/subscription/beb/reconciler_test.go:102]: How about `When("Updating the clean event types in the Subscription status"`?

File: components/eventing-controller/testing/safesubscriptions.go [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+
+import (
+	"strings"
+	"sync"
+
+	bebtypes "github.com/kyma-project/kyma/components/eventing-controller/pkg/ems/api/events/types"
+)
+
+// SafeSubscription encapsulates Subscriptions to provide mutual exclusion.
+type SafeSubscription struct {
```
Comment (marcobebway) [inline components/eventing-controller/testing/safesubscriptions.go:11]: Please change `SafeSubscription` to `SafeSubscriptions`.

File: components/eventing-controller/controllers/subscription/beb/reconciler_test.go [hunk 17]
```diff
@@ -896,11 +1004,7 @@ var _ = Describe("Subscription Reconciliation Tests", func() {
 
 			Context("Delete BEB Subscription", func() {
 				By("Deleting its entry in BEB mock internal cache", func() {
-					for k := range beb.Subscriptions {
-						if strings.Contains(k, nameMapper.MapSubscriptionName(kymaSubscription)) {
-							delete(beb.Subscriptions, k)
-						}
-					}
+					beb.Subscriptions.DeleteSubscriptionsByName(nameMapper.MapSubscriptionName(kymaSubscription))
```
Comment (marcobebway) [inline components/eventing-controller/controllers/subscription/beb/reconciler_test.go:1007]: I think renaming `beb` to `bebMock` would improve the readability.

File: components/eventing-controller/controllers/subscription/beb/reconciler_test.go [hunk 18]
```diff
@@ -1072,9 +1176,16 @@ func ensureSubscriptionCreated(ctx context.Context, subscription *eventingv1alph
 	Expect(err).Should(BeNil())
 }
 
+// ensureSubscriptionCreated creates a Subscription in the k8s cluster. If a custom namespace is used, it will be created as well.
```
Comment (marcobebway) [inline components/eventing-controller/controllers/subscription/beb/reconciler_test.go:1179]: Should be `ensureSubscriptionUpdated` as well.

File: components/eventing-controller/pkg/handlers/nats.go [hunk 19]
```diff
@@ -121,18 +125,23 @@ func (n *Nats) SyncSubscription(sub *eventingv1alpha1.Subscription, cleaner even
...
-				log.Errorw("create NATS subscription failed", "error", err)
+				log.Errorw("creating NATS subscription failed", "error", subscribeErr)
 				return false, subscribeErr
 			}
 			n.subscriptions[createKey(sub, subject, i)] = natsSub
 		}
 	}
-	sub.Status.Config = subsConfig
+
+	if len(cleanSubjects) > 0 {
```
Comment (marcobebway) [inline components/eventing-controller/pkg/handlers/nats.go:140]: Why do we need this check? why not always overwrite the `sub.Status.CleanEventTypes` with the cleaned ones?

File: components/eventing-controller/testing/bebmock.go [hunk 20]
```diff
@@ -98,12 +96,14 @@ func (m *BEBMock) Start() string {
...
 			return
 		}
+
 		// messaging API request
 		if strings.HasPrefix(r.RequestURI, MessagingURLPath) {
 			switch r.Method {
 			case http.MethodDelete:
 				key := r.URL.Path
-				delete(m.Subscriptions, key)
+				//delete(m.Subscriptions, key) //todo
```
Comment (marcobebway) [inline components/eventing-controller/testing/bebmock.go:105]: Do we still need this todo?

File: components/eventing-controller/testing/test_helpers.go [hunk 21]
```diff
@@ -269,22 +269,32 @@ func WithNotCleanEventTypeFilter(s *eventingv1alpha1.Subscription) {
 // WithFilter sets the Subscription filter with the given event source and type.
```
Comment (marcobebway) [inline components/eventing-controller/testing/test_helpers.go:269]: Comment is not accurate anymore, this function now adds one more filter to the existing ones.

File: components/eventing-controller/controllers/subscription/beb/reconciler_test.go (hunk 17 above)
Comment (FriedrichWilken) [reply components/eventing-controller/controllers/subscription/beb/reconciler_test.go:1007]: I considered this too but in this scope there is another variable dubbed `bebMock`. I will rename 
`bebMock` to `mock` and
`beb` to `bebMock`.
What do you think?

**Review by marcobebway — DISMISSED** [review]

**Review by marcobebway — APPROVED** [review]
/lgtm

**Review by marcobebway — APPROVED** [review]
